source/common/network/filter_matcher.cc

```diff
@@ -0,0 +1,186 @@
+      if (matchers_[i]->matches(cb)) {
+        return true;
+      }
+    }
+    return false;
+  }
+  void
+  complete(const envoy::config::listener::v3::ListenerFilterChainMatchPredicate& self) override {
+    sub_matcher_offset_ = matchers_.size();
+    sub_matcher_len_ = self.or_match().rules_size();
+    for (const auto& sub : self.or_match().rules()) {
+      matchers_.emplace_back(createFromMessage(sub, matchers_));
+    }
+    TwoPhaseMatcher* matcher{nullptr};
+    for (uint32_t i = 0; i < sub_matcher_len_; i++) {
+      matcher = static_cast<TwoPhaseMatcher*>(matchers_[i + sub_matcher_offset_].get());
+      matcher->complete(self.or_match().rules(i));
+    }
+  }
+
+private:
+  std::vector<ListenerFilterMatcherPtr>& matchers_;
+  uint32_t sub_matcher_offset_{0};
+  uint32_t sub_matcher_len_{0};
+};
+
+struct DstPortMatcher final : public TwoPhaseMatcher {
+public:
+  DstPortMatcher() = default;
+  bool matches(ListenerFilterCallbacks& cb) const override {
+    const auto& address = cb.socket().localAddress();
+    // Match on destination port (only for IP addresses).
+    if (address->type() == Address::Type::Ip) {
+      const auto port = address->ip()->port();
+      return start_ <= port && port < end_;
+    } else {
+      return true;
+    }
+  }
+  void
+  complete(const envoy::config::listener::v3::ListenerFilterChainMatchPredicate& self) override {
+    ASSERT(self.rule_case() == envoy::config::listener::v3::ListenerFilterChainMatchPredicate::
+                                   RuleCase::kDestinationPortRange);
+    start_ = self.destination_port_range().start();
+    end_ = self.destination_port_range().end();
+  }
+
+private:
+  uint32_t start_;
+  uint32_t end_;
+};
+namespace {
+std::unique_ptr<TwoPhaseMatcher> createFromMessage(
+    const envoy::config::listener::v3::ListenerFilterChainMatchPredicate& match_config,
+    std::vector<ListenerFilterMatcherPtr>& matchers) {
+  switch (match_config.rule_case()) {
+  case envoy::config::listener::v3::ListenerFilterChainMatchPredicate::RuleCase::kAnyMatch:
+    return std::make_unique<TrueMatcher>();
+  case envoy::config::listener::v3::ListenerFilterChainMatchPredicate::RuleCase::kNotMatch: {
+    return std::make_unique<NotMatcher>(matchers);
+  }
+  case envoy::config::listener::v3::ListenerFilterChainMatchPredicate::RuleCase::kAndMatch: {
+    return std::make_unique<AndMatcher>(matchers);
+  }
+  case envoy::config::listener::v3::ListenerFilterChainMatchPredicate::RuleCase::kOrMatch: {
+    return std::make_unique<OrMatcher>(matchers);
+  }
+  case envoy::config::listener::v3::ListenerFilterChainMatchPredicate::RuleCase::
+      kDestinationPortRange: {
+    return std::make_unique<DstPortMatcher>();
+  }
+  // Invalid message.
+  case envoy::config::listener::v3::ListenerFilterChainMatchPredicate::RuleCase::RULE_NOT_SET: {
+    throw EnvoyException(
+        absl::StrFormat("invalid listener filter chain matcher: %s", match_config.DebugString()));
+  }
```

Comment:
This should be impossible via schema (make sure there is a validator). Same below. Use NOT_REACHED instead

Reply:
Sure. Will add validator if it doesn't exist yet.

Reply:
I have different opinions on NOT_REACHED: 
In case if the control plane api has newer version so a new case is provided, NOT_REACHED is hit and the envoy is terminate I suppose.
Throw an exception here allows envoy rejecting the listener config. 
Do you prefer terminate the envoy to received the signal in the aggressive crash way, or assuming envoy will be updated soon while leave the envoy running under previous listener config?

Reply:
If you have a validator it will throw an exception and the config will be rejected. This is what we do everywhere else.

Reply:
Thanks! I am aware of something like allow-unknown-field but I don't know how could the validator work while we do allow unknown field in this example. 
Could you help me understand if it's a particular validation rule to achieve this, or it's envoy's  default behavior to reject such a config as a-new-matcher in this example?
Meanwhile following your comment to fix

Reply:
`option (validate.required) = true;` does it which you have.

Reply:
Thanks! I get your point. I didn't know `required` also check if it's *unknown value* ...
Thank you for the information!